client/tso_stream.go

```diff
@@ -172,51 +178,252 @@ func (s tsoTSOStreamAdapter) Recv() (tsoRequestResult, error) {
 	}, nil
 }
 
+type onFinishedCallback func(result tsoRequestResult, reqKeyspaceGroupID uint32, err error)
+
+type batchedRequests struct {
+	startTime          time.Time
+	count              int64
+	reqKeyspaceGroupID uint32
+	callback           onFinishedCallback
+}
+
+// tsoStream represents an abstracted stream for requesting TSO.
+// This type designed decoupled with users of this type, so tsoDispatcher won't be directly accessed here.
+// Also in order to avoid potential memory allocations that might happen when passing closures as the callback,
+// we instead use the `batchedRequestsNotifier` as the abstraction, and accepts generic type instead of dynamic interface
+// type.
 type tsoStream struct {
 	serverURL string
 	// The internal gRPC stream.
 	//   - `pdpb.PD_TsoClient` for a leader/follower in the PD cluster.
 	//   - `tsopb.TSO_TsoClient` for a primary/secondary in the TSO cluster.
 	stream grpcTSOStreamAdapter
+	// An identifier of the tsoStream object for metrics reporting and diagnosing.
+	streamID string
```

Comment:
Is it possible to attach this ID to each TSO request so we can know which stream a TSO request is dispatched to and put some debug info around it as well?

Reply:
I added it to the `tsoRequest` but it's for now not used. Do have any idea about how to add the new debug info you mentioned?

Reply:
I was thinking about whether we could add fields to `TsoRequest`, so that the client side could also get the stream ID that processes its own TSO request, but considering that we would need to modify kvproto, it's okay if we don't do this in this PR.